templates/extras/simp.md

```diff
@@ -1,111 +1,254 @@
 # Simp
 
-The `simp` tactic works by applying a conditional term rewriting system to try and prove, or at
-least simplify, your goal. What this basically means is that `simp` is equipped with a list of
-lemmas (those tagged with the `simp` attribute), many of which are of the form `X = Y` or `X iff Y`,
-and attempts to match subterms of the goal with the left hand side of a rule, and then replaces the
-subterm with the right hand side. The system is conditional in the sense that lemmas are allowed to
-have preconditions (`P -> (X = Y)`) and in these cases it will try and prove the precondition using
-its simp lemmas before applying `X = Y`.
+## Overview
 
-You can watch `simp` in action by using `set_option trace.simplify true` in your code. For example
+In this document we will explain basic usage of Lean's simplifier,
+`simp`, and the related tactic `dsimp`.
+
+## Introduction
+
+Lean has a "simplifier", called `simp`. This is a tactic whose job,
+unfortunately for beginners, is *not* to prove an arbitrary simple
+statement. Lean's simplifier is what is called a "conditional term
+rewriting system". Rewriting is the act of changing a goal of the form
+`P(A)` into `P(B)`, given a hypothesis of the form `A = B` or `A ↔
+B`. The simplifier's job is to try and prove, or at least simplify,
+goals or hypotheses, by using repeated rewrites. Here is an example
+(using `mathlib`).
 
 ```lean
-namespace hidden
+import algebra.group.defs
 
-definition cong (a b m : ℤ) : Prop := ∃ n : ℤ, b - a = m * n
+variables (G : Type) [group G] (a b c : G)
 
-notation a ` ≡ ` b ` mod ` m  := cong a b m
-set_option trace.simplify true
-theorem cong_refl (m : ℤ) : ∀ a : ℤ, a ≡ a mod m :=
+example : a * a⁻¹ * 1 * b = b * c * c⁻¹ :=
 begin
-intro a,
-unfold cong,
-existsi (0:ℤ),
-simp
+  simp
 end
-
-end hidden
 ```
 
-If you do this exercise you will discover firstly that `simp` spends a lot of its time trying random
-lemmas and then giving up very shortly afterwards, and also that the `unfold` command is also
-underlined in green -- Lean seems to apply `simp` when you do an `unfold` as well (apparently
-`unfold` just asks `simp` to do its dirty work for it -- `unfold X` is close to `simp only [X]`).
-
-If you only want to see what worked rather than all the things that didn't, you could try
-`set_option trace.simplify.rewrite true`.
+How would a human solve that goal? They would notice that `a * a⁻¹ =
+1`, that `1 * 1 = 1` and so on, until they had simplified the example
+to `b = b`, which is obviously true. This is also what the simplifier
+is doing. Indeed, if you add `set_option trace.simplify.rewrite true`
```

Comment:
Could you mention `set_option trace.simplify true` as well? I think I've referred a few people to this page on Zulip for these trace options.

Reply:
Done.